cpp/src/arrow/compute/kernels/nth_to_indices.cc

```diff
@@ -0,0 +1,205 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/kernels/nth_to_indices.h"
+
+#include <algorithm>
+#include <numeric>
+#include <vector>
+
+#include "arrow/builder.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/expression.h"
+#include "arrow/compute/logical_type.h"
+#include "arrow/type_traits.h"
+
+namespace arrow {
+
+class Array;
+
+namespace compute {
+
+/// \brief Kernel implementing NthToIndices operation
+class ARROW_EXPORT NthToIndicesKernel : public OpKernel {
+ protected:
+  std::shared_ptr<DataType> type_;
+
+ public:
+  /// \brief Kernel interface
+  ///
+  /// delegates to subclasses via NthToIndices()
+  virtual Status Call(FunctionContext* ctx, const Datum& values, uint64_t n,
+                      Datum* offsets) = 0;
+
+  /// \brief output type of this kernel
+  std::shared_ptr<DataType> out_type() const override { return uint64(); }
+
+  /// \brief single-array implementation
+  virtual Status NthToIndices(FunctionContext* ctx, const std::shared_ptr<Array>& values,
+                              uint64_t n, std::shared_ptr<Array>* offsets) = 0;
+
+  /// \brief factory for NthToIndicesKernel
+  ///
+  /// \param[in] value_type constructed NthToIndicesKernel will support
+  ///            values of this type
+  /// \param[out] out created kernel
+  static Status Make(const std::shared_ptr<DataType>& value_type,
+                     std::unique_ptr<NthToIndicesKernel>* out);
+};
+
+template <typename ArrayType>
+bool CompareValues(const ArrayType& array, uint64_t lhs, uint64_t rhs) {
+  return array.Value(lhs) < array.Value(rhs);
```

Comment:
Note that `GetView` also exists on numeric arrays, so you don't need two different functions.

Reply:
Thanks. It simplifies code significantly. I think sorting kernel should also be refined with it.